### PR TITLE
Improvements for ElasticsearchVectorStore

### LINKED ISSUE #834: [Feature] Improvements in ElasticVectorStore
Hey, here are some proposals on improvements in ElasticVectorStore. I have forked the project already and made some basic changes to fit ElasticVectorStore to my needs. You can see the commits in here: 
https://github.com/hwchase17/langchain/compare/master...yasha-dev1:langchain:master
Still I will have to write unit tests for it and test it more thoroughly. 
However these are the implementations that I have already done:
1. Added https connection and basic Auth for connection to elasticsearch. I have to Add `ElasticConf` in order to incorporate Auth settings into the library because Elasticsearch is suppoused to be an optional dep
2. Added Filter for all types of `VectorStores` due to the fact that for large scale data, Ann would be really slow and in most of my use cases, I need to fetch the similarity in a filtered context. I have added filter for the metadata that gets saved on to vector store in order to filter the data before performing similarity search. (Its already implemented by most of the vector stores, so its as easy as changing the query)
3. Added `setup_index` in order for the client, to specify the schema structure of the data saved there (aka metadata). For example in my use case, I didn't want to index some specific metadata in elasticsearch, that's why I needed to specify my own data mapping. But also some mapping logic depends on the lib, so best would be that the client passes the metadata schema structure to each `VectorStore` so that they get created if it doesn't exist.

_Note: All of the above suggested features are already implemented in the mentioned commits, However I will have to do more thourough testing. And It would be nice to get a feedback if this doesn't meet some of your lib's rules, Because the overall interface of `VectorStore` has been changed significantly. these are only my needs that I implemented, But it would be cool for you guys to review, and potentially, if I have done more tests, I can make a PR_

### PR DESCRIPTION
**Pull request by yasha-dev1**
- Added `setup_index` to the interface `VectorStore` for the procedure of creating and setting up the data schema 
- Added `DataSchemaBuilder` as a base interface for all `vectorStores`
- Added a special `data_schema_builder` for elasticsearch for creating index mapping in elasticsearch
- Added `ElasticConf`. A context to be passed to `ElasticVectorStore` to pass the credential context to the object (Not tested properly yet)
- Added `similarity_search_by_id` and `similarity_search_by_vector`, this basically seperates the logic in `similar_search` more accurately in methods
- Added `VectorStoreFilter`, a functionality to filter the metadata before ANN indexing
- incorporated query filter instead of default elasticsearch filter
- implemented concrete class `ElasticFilter`

close #834

### CODE REVIEW

**Comment by yasha-dev1** [comment]
Still I haven't added code for unit tests. I will add, once the mentioned changes for the `VectorStore` itnerface are approved

File: langchain/vectorstores/filters/base.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+"""Interface for vector store Filters."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Optional
+
+
+class VectorStoreFilter(ABC):
```
Comment (hwchase17) [inline langchain/vectorstores/filters/base.py:8]: how much of this is actually generic vs just being needed for elastic search?

**Review by hwchase17 — COMMENTED** [review]
my main questions around this are basically are the changes to the base class actually generic, or are they specific to elastic vector store? even the title suggests they may be specific to elastic. if thats the case, im reluctant to change the base class

**Comment by yasha-dev1** [comment]
Hi, 
To answer the first question regarding how much generic are the operations done in base `VectorStore`:
I went through the documentation of all subclasses of `VectorStore`. 
in this PR I have added `setup_index`, `similarity_search_by_id`, `similarity_search_by_vector` to the base class, so the new functionalies are `creating an index or collection that represents a set of vectors`, `querying based on document ID`, `querying based on the vector`. so If all the subclass vector stores are able to do that, then we can add it to base class.

And here's what I found out of each different datastores:


**Weaviate**
- [supports indexes](https://weaviate.io/developers/weaviate/configuration/schema-configuration#class-object), I believe the terminology is called `Class object`
- [supports querying based on ID](https://weaviate.io/developers/weaviate/api/graphql/get#vector-search-operators)
- [supports querying based on vector](https://weaviate.io/developers/weaviate/api/graphql/vector-search-parameters#built-in-parameters)

**Qdrant**
- [supports indexes](https://qdrant.github.io/qdrant/redoc/index.html), I believe the terminology is called `Collection`
- [supports querying based on ID](https://qdrant.github.io/qdrant/redoc/index.html#tag/points/operation/get_point)
- [supports querying based on vector](https://qdrant.github.io/qdrant/redoc/index.html#tag/points/operation/search_points)

**Pinecone**
- [supports indexes](https://docs.pinecone.io/docs/indexes)
- supports querying based on ID, Pinecone doesn't have ID foreach document, but it can be stored in metadata and generated from langchain if no doc_id is provided
- [supports querying based on vector](https://docs.pinecone.io/docs/query-data#using-metadata-filters-in-queries)

**Qdrant**
- [supports indexes](https://milvus.io/docs/create_collection.md#Create-a-Collection), I believe the terminology is called `Collection`
- supports querying based on ID, again Milvus doesn't have ID foreach document, but it can be stored in metadata and generated from langchain if no doc_id is provided
- [supports querying based on vector](https://milvus.io/docs/search.md#Conduct-a-vector-search) using data field in query time

As for Faiss, I'm not sure, haven't checked it thoroughly, there is a concept of index but I don't think you can actually filter the data or search by metadata.
please correct me on above data, but it seems most of the Vector stores already support the implemented functionality

File: langchain/vectorstores/filters/base.py (hunk 1 above)
Comment (yasha-dev1) [reply langchain/vectorstores/filters/base.py:8]: As for this, This is basically a FilterBuilder, to build filters for a specific datastore. As I went through all the vectorStores, they actually support filtering before ANN search (except FAISS, not sure)
The idea of this interface is only to build a filter, for different types of vectorstores and then build the final query

**Comment by kacperlukawski** [comment]
@yasha-dev1 I like the idea of filtering within vector stores! I can help with Qdrant implementation. I'm wondering how we could build some more complex queries (even thinking about simple AND/OR statements). 

I'd say the `VectorStoreFilter` should be common for all the engines, so from the user perspective there is no need to choose a different class if they decide to switch to a different provider. I imagine the `VectorStoreFilter` works as some sort of tree-like structure to keep all the criteria, and then they got parsed into the engine-specific structure. Either by a particular `VectorStore` or even better by some sort of engine parser, but definitely not in the `VectorStoreFilter` directly.

**Comment by yasha-dev1** [comment]
Yes, `VectorStoreFilter` should be a general interface for all vector stores. The best solution which comes to my mind is to implement ES-like Interface. for now only `And` operator is supported, but we can work on making it more general with `OR` operator. sth like ES [bool query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html)
Collaborations are more than welcome, I can add you as collaborator

**Comment by SirBernardPhilip** [comment]
I too would like this functionality, is there any way I can help @yasha-dev1?

**Pull request closed without merging**